src/main/java/com/vmware/vipclient/i18n/messages/api/opt/local/LocalLocaleOpt.java

```diff
@@ -37,72 +39,104 @@ public LocalLocaleOpt(LocaleDTO dto) {
-				return supportedLanguageNames;
+				return;
 			for(String language : supportedLanguages){
 				if(!ConstantsKeys.SOURCE.equalsIgnoreCase(language))
 					supportedLanguageNames.put(language, (String) languagesNames.get(language));
 			}
 		}
-		return supportedLanguageNames;
+		if (!supportedLanguageNames.isEmpty()) {
+			cacheItem.addCachedData(supportedLanguageNames);
+			cacheItem.setTimestamp(System.currentTimeMillis());
+		}
     }
 
     private Map<String, String> getLanguagesNamesFromCLDR(String locale){
-		Map<String, String> languagesNames = null;
+		LocaleCacheItem languagesNames = null;
 		logger.debug("Look for languages' names from cache for locale [{}]!", locale);
 		FormattingCacheService formattingCacheService = new FormattingCacheService();
 		languagesNames = formattingCacheService.getLanguagesNames(locale);// key
 		if (languagesNames != null) {
+			if (languagesNames.isExpired()) { // cacheItem has expired
+				// Update the cache in a separate thread
+				populateLanguagesCache(locale, languagesNames);
+			}
 			logger.debug("Find languages' names from cache for locale [{}]!", locale);
-			return languagesNames;
+			return languagesNames.getCachedData();
 		}
-		languagesNames = getLanguagesNamesFromBundle(locale);
-		if (languagesNames != null) {
+		languagesNames = new LocaleCacheItem();
+		getLanguagesNamesFromBundle(locale, languagesNames);
+		if (!languagesNames.getCachedData().isEmpty()) {
 			logger.debug("Find the languages' names from local bundle for locale [{}].\n", locale);// [datetime] and
 			formattingCacheService.addLanguagesNames(locale, languagesNames);
 			logger.debug("Languages' names are cached for locale [{}]!\n\n", locale);
-			return languagesNames;
 		}
-		return null;
+		return languagesNames.getCachedData();
 	}
 
-	private JSONObject getLanguagesNamesFromBundle(String locale) {
+	private void getLanguagesNamesFromBundle(String locale, LocaleCacheItem cacheItem) {
 		logger.debug("Look for languages' names from local package bundle for locale [{}]!", locale);
 		String normalizedLocale = CommonUtil.getCLDRLocale(locale, localePathMap, localeAliasesMap);
 		logger.info("Normalized locale for locale [{}] is [{}]", locale, normalizedLocale);
 		if(normalizedLocale == null || normalizedLocale.isEmpty())
-			return null;
-        try {
-            String languagesJsonStr = PatternUtil.getLanguageFromLib(normalizedLocale);
-			JSONObject languagesData = (JSONObject) new JSONParser().parse(languagesJsonStr);
-            return (JSONObject) languagesData.get(PatternKeys.LANGUAGES);
+			return;
+		try {
+			String languagesJsonStr = PatternUtil.getLanguageFromLib(normalizedLocale);
+			Map<String, Object> languagesData = (Map<String, Object>) new JSONParser().parse(languagesJsonStr);
+			if (languagesData != null) {
+				cacheItem.addCachedData((Map<String, String>) languagesData.get(PatternKeys.LANGUAGES));
+				cacheItem.setTimestamp(System.currentTimeMillis());
```

Comment:
To avoid mismatched etag, timestamp and maxAgeMillis due to thread unsafety, have to set all the factors that go together (etag, timestamp and maxAgeMillis, and cachedData) in one synchronized setter method. 
So can't move set timestamp to service layer.